src/System.Private.CoreLib/shared/System/Text/ASCIIEncoding.cs

```diff
@@ -357,564 +398,511 @@ public override unsafe string GetString(byte[] bytes, int byteIndex, int byteCou
         //
 
-        // GetByteCount
-        // Note: We start by assuming that the output will be the same as count.  Having
-        // an encoder or fallback may change that assumption
-        internal sealed override unsafe int GetByteCount(char* chars, int charCount, EncoderNLS encoder)
+        internal sealed override unsafe int GetByteCount(char* pChars, int charCount, EncoderNLS encoder)
         {
-            // Just need to ASSERT, this is called by something else internal that checked parameters already
-            Debug.Assert(charCount >= 0, "[ASCIIEncoding.GetByteCount]count is negative");
-            Debug.Assert(chars != null, "[ASCIIEncoding.GetByteCount]chars is null");
+            Debug.Assert(charCount >= 0);
+            Debug.Assert(charCount == 0 || pChars != null);
 
-            // Assert because we shouldn't be able to have a null encoder.
-            Debug.Assert(encoderFallback != null, "[ASCIIEncoding.GetByteCount]Attempting to use null fallback encoder");
+            // We treat the incoming data as UCS-2, _not_ UTF-16. This means that if the incoming data
+            // has UTF-16 surrogate code points (whether standalone or well-formed in pairs), each code
+            // point is run through the fallback buffer in isolation. This allows us to perform an
+            // optimization: if the fallback buffer's behavior is to replace any non-ASCII UCS-2 chars
+            // with a single ASCII char, we can just return the input length and be done with it.
 
-            char charLeftOver = (char)0;
-            EncoderReplacementFallback fallback = null;
+            int indexOfFirstNonAsciiChar = charCount;
 
-            // Start by assuming default count, then +/- for fallback characters
-            char* charEnd = chars + charCount;
+            // The <= 0 instead of the == 0 is so that the JIT can elide the 'length' parameter check
+            // in the ROS ctor just in case it needs to be called at the end of this method.
 
-            // For fallback we may need a fallback buffer, we know we aren't default fallback.
-            EncoderFallbackBuffer fallbackBuffer = null;
-            char* charsForFallback;
-
-            if (encoder != null)
+            if ((charCount <= 0)
+                || (((encoder == null) ? this.EncoderFallback : encoder.Fallback) is EncoderReplacementFallback replacementFallback && replacementFallback.MaxCharCount == 1 && replacementFallback.DefaultString[0] <= 0x007Fu))
             {
-                charLeftOver = encoder._charLeftOver;
-                Debug.Assert(charLeftOver == 0 || char.IsHighSurrogate(charLeftOver),
-                    "[ASCIIEncoding.GetByteCount]leftover character should be high surrogate");
+                goto Finished; // no work to be done
+            }
 
-                fallback = encoder.Fallback as EncoderReplacementFallback;
+            // If we got this far, we have complex fallback semantics.
```

Comment:
Do we need to give EncoderFallback a chance to run first - in case it has something cached from before?

Reply:
Yes, I missed this in the original replumbing and am trying to figure out how to work it back in without interfering with the fast path logic. My ideal scenario is that this logic would just be moved to `EncoderNLS` so that it can be shared across `ASCIIEncoding` / `UTF8Encoding` / (all the other encoding types) because it's going to be the same across all of them: (a) see if there's a remaining high surrogate from the previous operation, (b) fold it back in with the next char to generate a full scalar value, (c) run that scalar value through the new `Encoding.GetBytes` or `Encoding.GetByteCount` routines which accept `Rune` as an input parameter, then (d) run the main `GetBytes(..., EncoderNLS)` / `GetByteCount(..., EncoderNLS)` logic. There's no reason this logic needs to be duplicated in every one of the `Encoding` derived classes.